# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/notification/NotificationParent.cpp
# Commit: f318fe78ff14
# Full Hash: f318fe78ff1444fe0c3c46495f749720c9c53c16
# Author: Kagami Sascha Rosylight <saschanaz@outlook.com>
# Date: 2024-11-08 04:19:40
# Regressor Bug: 1928702
# File Overlap Count: 1
# Description:
#   Bug 1928702 - Part 7: Settle ShowResolver r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D227692
# ==============================================================================

diff -r 3daeacd08336 -r f318fe78ff14 dom/notification/NotificationParent.cpp
--- a/dom/notification/NotificationParent.cpp	Thu Nov 07 17:32:05 2024 +0000
+++ b/dom/notification/NotificationParent.cpp	Thu Nov 07 17:32:06 2024 +0000
@@ -28,10 +28,30 @@
   if (!strcmp("alertshow", aTopic)) {
     (void)NS_WARN_IF(NS_FAILED(
         AdjustPushQuota(mPrincipal, NotificationStatusChange::Shown)));
+    if (!mResolver) {
+#ifdef ANDROID
+      // XXX: This can happen as we resolve showNotification() immediately on
+      // Android for now and a mock service may still call this.
+      return NS_OK;
+#else
+      MOZ_ASSERT_UNREACHABLE("Are we getting double show events?");
+      return NS_ERROR_FAILURE;
+#endif
+    }
+    mResolver.take().value()(CopyableErrorResult());
+    return NS_OK;
   }
   if (!strcmp("alertfinished", aTopic)) {
     (void)NS_WARN_IF(NS_FAILED(
         AdjustPushQuota(mPrincipal, NotificationStatusChange::Closed)));
+    if (mResolver) {
+      // alertshow happens first before alertfinished, and it should have If
+      // not it means it failed to show and is bailing out.
+      // XXX: Apparently XUL manual do not disturb mode does this without firing
+      // alertshow at all.
+      mResolver.take().value()(CopyableErrorResult(NS_ERROR_FAILURE));
+      return NS_OK;
+    }
   }
   return NS_OK;
 }
@@ -39,10 +59,17 @@
 // Step 4 of
 // https://notifications.spec.whatwg.org/#dom-notification-notification
 mozilla::ipc::IPCResult NotificationParent::RecvShow(ShowResolver&& aResolver) {
+  mResolver.emplace(std::move(aResolver));
+
   // Step 4.2: Run the fetch steps for notification. (Will happen in
   // nsIAlertNotification::LoadImage)
   // Step 4.3: Run the show steps for notification.
-  Show();
+  nsresult rv = Show();
+  if (NS_FAILED(rv)) {
+    mResolver.take().value()(CopyableErrorResult(rv));
+  }
+  // If not failed, the resolver will be called asynchronously by
+  // NotificationObserver
   return IPC_OK();
 }
 
@@ -107,6 +134,15 @@
   MOZ_TRY(alertService->ShowPersistentNotification(
       NS_ConvertUTF8toUTF16(persistentData.StringCRef()), alert, this));
 
+#ifdef ANDROID
+  // XXX: the Android nsIAlertsService is broken and doesn't send alertshow
+  // properly, which means we cannot depend on it to resolve the promise. For
+  // now we resolve the promise here.
+  // (This now fires onshow event regardless of the actual result, but it should
+  // be better than the previous behavior that did not do anything at all)
+  mResolver.take().value()(CopyableErrorResult());
+#endif
+
   return NS_OK;
 }
 